src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,712 @@
+        {
+            return o << s.span << "(T:" << s.tipSupport
+                     << ",B:" << s.branchSupport << ")";
+        }
+
+        Json::Value
+        getJson() const
+        {
+            Json::Value res;
+            res["id"] = to_string(span.tip().second);
+            res["seq"] = static_cast<std::uint32_t>(span.tip().first);
+            res["tipSupport"] = tipSupport;
+            res["branchSupport"] = branchSupport;
+            if(!children.empty())
+            {
+                Json::Value &cs = (res["children"] = Json::arrayValue);
+                for (auto const& child : children)
+                {
+                    cs.append(child->getJson());
+                }
+            }
+            return res;
+        }
+    };
+
+    // The root of the trie. The root is allowed to break the no-single child
+    // invariant.
+    std::unique_ptr<Node> root;
+
+    /** Find the node in the trie that represents the longest common ancestry
+        with the given ledger.
+
+        @return Pair of the found node and the sequence number of the first
+                ledger difference.
+    */
+    std::pair<Node*, Seq>
+    find(Ledger const& ledger) const
+    {
+        Node* curr = root.get();
+
+        // Root is always defined and is in common with all ledgers
+        assert(curr);
+        Seq pos = curr->span.diff(ledger);
+
+        bool done = false;
+
+        // Continue searching for a better span as long as the current position
+        // matches the entire span
+        while (!done && pos == curr->span.end())
+        {
+            done = true;
+            // Find the child with the longest ancestry match
+            for (std::unique_ptr<Node> const& child : curr->children)
+            {
+                auto const childPos = child->span.diff(ledger);
+                if (childPos > pos)
+                {
+                    done = false;
+                    pos = childPos;
+                    curr = child.get();
+                    break;
+                }
+            }
+        }
+        return std::make_pair(curr, pos);
+    }
+
+    void
+    dumpImpl(std::ostream& o, std::unique_ptr<Node> const& curr, int offset)
+        const
+    {
+        if (curr)
+        {
+            if (offset > 0)
+                o << std::setw(offset) << "|-";
+
+            std::stringstream ss;
+            ss << *curr;
+            o << ss.str() << std::endl;
+            for (std::unique_ptr<Node> const& child : curr->children)
+                dumpImpl(o, child, offset + 1 + ss.str().size() + 2);
+        }
+    }
+
+public:
+    LedgerTrie() : root{std::make_unique<Node>()}
+    {
+    }
+
+    /** Insert and/or increment the support for the given ledger.
+
+        @param ledger A ledger and its ancestry
+        @param count The count of support for this ledger
+     */
+    void
+    insert(Ledger const& ledger, std::uint32_t count = 1)
+    {
+        Node* loc;
```

Comment:
What does `loc` stand for?

Reply:
`loc` is the location where the location where the new ledger support will go; it might be a `Node` that matches the ledger exactly, or it might be the ancestor.

Any alternate name that you would suggest?

Reply:
Got it. That's fine. Sometimes I'll put a comment next to a variable like that and keep the small name: `Node* loc; // location of the new ledger support`, but comments like that can clutter up the code too. I think it's fine as-is, I just didn't think of "location". Thanks for clarifying.